fallback.go

```diff
@@ -36,11 +42,16 @@ func (f fallbackProxyCreator) New(trace plugin.Handler, upstream proxy.Upstream)
 }
 
 func New(trace plugin.Handler) (f *Fallback) {
-	return &Fallback{trace: trace, rules: make(map[int]proxy.Upstream), proxy: fallbackProxyCreator{}}
+	return &Fallback{trace: trace, rules: make(map[int]rule), proxy: fallbackProxyCreator{}}
 }
 
 // ServeDNS implements the plugin.Handler interface.
 func (f Fallback) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	// If fallback has original option set for any code then copy original request to use it instead of changed
+	var originalRequest *dns.Msg
+	if f.original {
```

Comment:
This is global flag which says if we have at least one ORIGINAL option set. Needed for performance improvement.